### PR TITLE
[ONNX] Added support for negative axes.

### PR DESCRIPTION
**Pull request by Ewa21**

### CODE REVIEW

File: src/ngraph/frontend/onnx_import/op/hardmax.cpp [hunk 1]
```diff
@@ -34,13 +34,11 @@ namespace ngraph
                     const auto input = node.get_ng_inputs().at(0);
                     const auto& input_shape = input->get_shape();
                     auto axis = node.get_attribute_value<std::int64_t>("axis", 1);
-
-                    ASSERT_VALID_ARGUMENT(node, axis >= 0 && axis < input_shape.size())
-                        << "The provided axis value " << axis
-                        << " does not match the input tensor dimensions";
+                    auto valid_axis =
+                        common::convert_negative_axis(axis, input->get_shape().size());
```
Comment (tsocha) [inline src/ngraph/frontend/onnx_import/op/hardmax.cpp]:
```suggestion
                        common::convert_negative_axis(axis, input_shape.size());
```

File: src/ngraph/frontend/onnx_import/op/onehot.cpp [hunk 2]
```diff
@@ -50,15 +51,8 @@ namespace ngraph
...
-                    if (axis < 0)
-                    {
-                        axis += indices_shape.size() + 1;
-                    }
-
-                    ASSERT_VALID_ARGUMENT(node, (axis >= 0) && (axis <= indices_shape.size()))
-                        << "invalid 'axis' attribute: "
-                        << node.get_attribute_value<std::int64_t>("axis", -1);
+                    std::size_t valid_axis =
+                        common::convert_negative_axis(axis, inputs.at(0)->get_shape().size() + 1);
```
Comment (tsocha) [inline src/ngraph/frontend/onnx_import/op/onehot.cpp]:
```suggestion
                        common::convert_negative_axis(axis, indices_shape.size() + 1);
```

File: src/ngraph/frontend/onnx_import/utils/common.cpp [hunk 3]
```diff
@@ -46,6 +46,39 @@ namespace ngraph
                                        static_cast<onnx::TensorProto_DataType>(onnx_type)));
             }
 
+            std::size_t convert_negative_axis(std::int64_t axis, std::size_t tensor_rank)
+            {
+                if (axis >= 0)
+                {
+                    return std::min(static_cast<size_t>(axis), tensor_rank);
```
Comment (tsocha) [inline src/ngraph/frontend/onnx_import/utils/common.cpp]: I wonder if we should adjust too big axis or just throw an exception.
@postrational what do you think about it?

File: src/ngraph/frontend/onnx_import/utils/reduction.cpp [hunk 4]
```diff
@@ -51,10 +54,6 @@ namespace ngraph
 
                 auto reduction_axes = detail::get_reduction_axes(node);
 
-                ASSERT_VALID_ARGUMENT(node, reduction_axes.size() <= data_shape.size())
```
Comment (tsocha) [inline src/ngraph/frontend/onnx_import/utils/reduction.cpp]: I'm not sure if we should remove this check.

File: src/ngraph/frontend/onnx_import/utils/common.cpp [hunk 5]
```diff
@@ -46,6 +46,41 @@ namespace ngraph
                                        static_cast<onnx::TensorProto_DataType>(onnx_type)));
             }
 
+            std::size_t validate_axis(const ngraph::onnx_import::Node& node,
+                                      std::int64_t axis,
+                                      std::size_t tensor_rank)
+            {
+                NGRAPH_CHECK(llabs(axis) < tensor_rank,
+                             node.get_description(),
+                             "Parameter axis out of the tensor rank.");
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/common.cpp]: Please add provided axis value to the error message.

File: src/ngraph/frontend/onnx_import/op/flatten.cpp [hunk 6]
```diff
@@ -33,13 +33,10 @@ namespace ngraph
                     NodeVector inputs{node.get_ng_inputs()};
                     auto data = inputs.at(0);
                     auto axis = node.get_attribute_value<std::int64_t>("axis", 1);
+                    auto valid_axis =
+                        common::validate_axis(node, axis, data->get_shape().size() + 1);
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/flatten.cpp]: Why are you using `size() + 1` here?
If this is needed, it requires a line of documentation explaining it, or better yet a well named variable to hold this value: `data->get_shape().size() + 1`

File: src/ngraph/frontend/onnx_import/op/onehot.cpp [hunk 7]
```diff
@@ -51,15 +52,8 @@ namespace ngraph
...
-                    {
-                        axis += indices_shape.size() + 1;
-                    }
-
-                    ASSERT_VALID_ARGUMENT(
-                        node, (axis >= 0) && (axis <= static_cast<int64_t>(indices_shape.size())))
-                        << "invalid 'axis' attribute: "
-                        << node.get_attribute_value<std::int64_t>("axis", -1);
+                    std::size_t valid_axis =
+                        common::validate_axis(node, axis, indices_shape.size() + 1);
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/onehot.cpp]: Why are you using `size() + 1` here?

File: src/ngraph/frontend/onnx_import/op/flatten.cpp [hunk 8]
```diff
@@ -33,13 +33,10 @@ namespace ngraph
                     NodeVector inputs{node.get_ng_inputs()};
                     auto data = inputs.at(0);
                     auto axis = node.get_attribute_value<std::int64_t>("axis", 1);
+                    auto valid_axis =
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/op/flatten.cpp]: The name `valid_axis` is not a great name for `axis`. Please use reassign to `axis` instead throughout the PR.

File: src/ngraph/frontend/onnx_import/utils/common.cpp [hunk 9]
```diff
@@ -46,6 +46,41 @@ namespace ngraph
                                        static_cast<onnx::TensorProto_DataType>(onnx_type)));
             }
 
+            std::size_t validate_axis(const ngraph::onnx_import::Node& node,
+                                      std::int64_t axis,
+                                      std::size_t tensor_rank)
+            {
+                std::int64_t new_axis = axis;
+
+                if (new_axis < 0)
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/common.cpp]: You are not really checking any `new` axis here, but the provided axis value. Using the name `new_axis` is misleading in this check.

File: src/ngraph/frontend/onnx_import/utils/common.cpp [hunk 10]
```diff
@@ -46,6 +46,41 @@ namespace ngraph
...
+                {
+                    new_axis = new_axis + tensor_rank;
+                    NGRAPH_CHECK(new_axis >= 0,
+                                 node.get_description(),
+                                 "Parameter axis out of the tensor rank.");
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/common.cpp]: Please return the provided `axis` value in the error message.

File: src/ngraph/frontend/onnx_import/utils/common.cpp [hunk 11]
```diff
@@ -46,6 +46,41 @@ namespace ngraph
...
+                }
+
+                NGRAPH_CHECK(new_axis < tensor_rank,
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/common.cpp]: I think you can avoid using `NGRAPH_CHECK` twice in this function.

File: src/ngraph/frontend/onnx_import/utils/common.hpp [hunk 12]
```diff
@@ -67,28 +68,30 @@ namespace ngraph
...
-            /// \brief      Handle negative axis value.
+            /// \brief      Handle out of range axis.
             ///
+            /// \param[in]  node        The node with requested axis.
             /// \param[in]  axis        The requested axis value.
-            /// \param[in]  tensor_dim  The corresponding tensor dimensionality.
+            /// \param[in]  tensor_rank  The corresponding tensor rank.
             ///
-            /// \tparam     T           Provided axis value type.
+            /// \return     If negative axis, then return sum of tensor rank and axis.
```
Comment (postrational) [inline src/ngraph/frontend/onnx_import/utils/common.hpp]: What is the `sum of tensor rank and axis`? Please add a note that it's as if you were counting the axes from the end.

File: src/ngraph/frontend/onnx_import/op/flatten.cpp (hunk 8 above)
Comment (tomdol) [reply src/ngraph/frontend/onnx_import/op/flatten.cpp]: The axis attribute retrieved from onnx node is `int64_t`. The return type of `validate_axis()` is `size_t`. You have to cast from `size_t` to `int64_t` to avoid compilation warnings. While casting you can run into problems if a value held by `size_t` is larger than `int64_t` can fit though.

File: src/ngraph/frontend/onnx_import/op/onehot.cpp [hunk 13]
```diff
@@ -51,15 +52,13 @@ namespace ngraph
...
-                    if (axis < 0)
-                    {
-                        axis += indices_shape.size() + 1;
-                    }
-
-                    ASSERT_VALID_ARGUMENT(
-                        node, (axis >= 0) && (axis <= static_cast<int64_t>(indices_shape.size())))
-                        << "invalid 'axis' attribute: "
-                        << node.get_attribute_value<std::int64_t>("axis", -1);
+                    // Accepted range for axis is [-r-1, r] where r = rank(indices). Validate
```
Comment (tsocha) [inline src/ngraph/frontend/onnx_import/op/onehot.cpp]:
```suggestion
                    // Accepted range for axis is [-r, r] where r = rank(indices)+1. Validate
```

File: src/ngraph/frontend/onnx_import/op/unsqueeze.cpp [hunk 14]
```diff
@@ -17,6 +17,7 @@
 #include "ngraph/op/fused/unsqueeze.hpp"
 #include "ngraph/op/constant.hpp"
 #include "squeeze.hpp"
+#include "utils/common.hpp"
```
Comment (tsocha) [inline src/ngraph/frontend/onnx_import/op/unsqueeze.cpp]:
```suggestion
```

File: src/ngraph/frontend/onnx_import/utils/common.cpp [hunk 15]
```diff
@@ -46,6 +46,68 @@ namespace ngraph
                                        static_cast<onnx::TensorProto_DataType>(onnx_type)));
             }
 
+            std::size_t validate_axis(const ngraph::onnx_import::Node& node,
+                                      std::int64_t axis,
+                                      std::int64_t tensor_rank)
+            {
+                // Accepted range of value for axis is [-tensor_rank, tensor_rank-1].
+                NGRAPH_CHECK(((axis >= -tensor_rank) && (axis <= tensor_rank - 1)),
```
Comment (tsocha) [inline src/ngraph/frontend/onnx_import/utils/common.cpp]:
```suggestion
                NGRAPH_CHECK(((axis >= -tensor_rank) && (axis < tensor_rank)),
```

File: src/ngraph/graph_util.cpp [hunk 16]
```diff
@@ -152,14 +152,20 @@ void ngraph::replace_node(std::shared_ptr<Node> target, std::shared_ptr<Node> re
     {
         auto common_args = ngraph::find_common_args(target, replacement);
 
-        auto set_replacement_prov = [replacement](std::shared_ptr<Node> node) {
-            replacement->merge_provenance_tags_from(node);
+        std::set<string> removed_subgraph_tags;
```
Comment (tsocha) [inline src/ngraph/graph_util.cpp]: Why these changes are included in this PR?

File: src/ngraph/node.cpp [hunk 17]
```diff
@@ -338,6 +338,14 @@ void Node::add_provenance_tag(const std::string& tag)
     m_provenance_tags.insert(tag);
 }
 
+void Node::add_provenance_tags(const std::set<std::string>& tag_set)
```
Comment (tsocha) [inline src/ngraph/node.cpp]: Why these changes are included in this PR?

File: src/ngraph/node.hpp [hunk 18]
```diff
@@ -417,8 +417,9 @@ namespace ngraph
         /// Set device placement
         void set_placement_index(size_t placement);
 
-        const std::unordered_set<std::string>& get_provenance_tags() const;
+        const std::set<std::string>& get_provenance_tags() const;
         void add_provenance_tag(const std::string& tag);
+        void add_provenance_tags(const std::set<std::string>& tag_set);
```
Comment (tsocha) [inline src/ngraph/node.hpp]: Why these changes are included in this PR?

File: test/provenance.cpp [hunk 19]
```diff
@@ -29,7 +29,7 @@ using namespace std;
 using namespace ngraph;
 using ::testing::Return;
 
-using ProvSet = std::unordered_set<std::string>;
+using ProvSet = std::set<std::string>;
```
Comment (tsocha) [inline test/provenance.cpp]: Why these changes are included in this PR?

File: src/ngraph/frontend/onnx_import/op/onehot.cpp (hunk 13 above)
Comment (Ewa21) [reply src/ngraph/frontend/onnx_import/op/onehot.cpp]: I think the range should stay [-r-1, r].  `rank(indices)+1` makes the range too wide on the right.

Comment (tsocha) [reply src/ngraph/frontend/onnx_import/op/onehot.cpp]: Yea you are right.
But now I can see that -r-1 is too wide from the left.
axis = -3
rank = 2
valid_axis => -3 + 2 = -1 which is invalid value

Comment (Ewa21) [reply src/ngraph/frontend/onnx_import/op/onehot.cpp]: Yes, you'er right. The range isn't symmetric. The rank sent to `validation_axis` is rank+1 and the only reason why is increased is because of the wider range on the left. I want to keep comments compatible to the ONNX documentation.

File: src/ngraph/frontend/onnx_import/op/flatten.cpp (hunk 8 above)
Comment (postrational) [reply src/ngraph/frontend/onnx_import/op/flatten.cpp]: Why is the return type of `validate_axis()` `size_t`?
